### PR TITLE
fix(azure-devops): roll back work item effort when deleting a timelog

### PR DESCRIPTION
**Pull request by genesiscz**
## Why

`tools azure-devops timelog add` and `import` push Remaining/Completed after they create rows. `delete` only removed the time log. Delete plus re-add double-counted Completed (real hits: `#296936` 30→38, `#302066` 14/16→23.5/6.5).

Plain arithmetic cannot undo a create that clamped Remaining at 0. A 6.5h log on Remaining 0 would revert to Remaining 6.5 if we just added the hours back.

## What changed

- `updateWorkItemEffort` now takes signed minutes, clamps both Remaining and Completed at 0, and appends one journal record per successful write to `~/.genesis-tools/azure-devops/effort-journal.jsonl`.
- `add`, interactive add, and `import` pass the created `timeLogIds`.
- `delete` resolves `{workItemId, minutes}`, then:
  1. If the newest journal row covers only this id and the work item still shows the journalled *After* values, restore the exact *Before* pair (this is the clamp case).
  2. Otherwise apply arithmetic (`Completed -= hours`, `Remaining += hours`, floored at 0) and print why the restore is approximate (no journal, batch import, or drifted fields).
- Effort failure stays non-fatal (`TF401320` on locked/Closed items). The row still deletes. Order is read+plan → delete row → write fields.

New flags:

| Flag | Behaviour |
|---|---|
| `--no-effort` | Delete the row only |
| `--dry-run` | Print the planned Remaining/Completed transition; write nothing |
| `--yes` | Skip the confirm prompt (required without a TTY) |

Non-TTY scripts that used `echo y | tools azure-devops timelog delete <id>` must switch to `--yes`.

## Tests

`bun run test src/azure-devops/timelog-effort.test.ts src/azure-devops/lib/timelog/effort-restore.test.ts src/azure-devops/lib/timelog/delete-entry.test.ts` — 27 pass.

Covers: 2h exact restore, clamped 8h restore (3/0 not 8/0), pre-journal arithmetic, single-id delete from a 5-entry import, `--no-effort`, `--dry-run`, locked WI, confirm cancel, query resolve, unresolved id still deletes.

Pre-commit biome + tsgo passed.

## Not in this PR

- Upper bound of Completed against OriginalEstimate (called out in the handoff as a follow-up).
- Batch `delete --workitem --all-matching`. Sequential `--yes` deletes are documented instead.
- The `col-fe-report` skill warning was updated locally (`~/.claude/skills/col-fe-report/SKILL.md` §7). That file is not in this repo.

## Summary by CodeRabbit

* **New Features**
  * Added support for rolling back Remaining and Completed effort when deleting time logs.
  * Added `--dry-run`, `--yes`, and `--no-effort` deletion options.
  * Added journal-based effort restoration with warnings for approximate results.
  * Improved effort tracking when adding or importing time logs.
  * Added clearer deletion help text and usage examples.

* **Bug Fixes**
  * Improved handling of locked work items, missing data, cancelled confirmations, and multi-entry imports.

* **Tests**
  * Expanded coverage for deletion, restoration, journaling, and edge cases.

### CODE REVIEW

File: src/azure-devops/commands/timelog/delete.ts [hunk 1]
```diff
@@ -1,73 +1,135 @@
...
+                    confirm:
+                        options.yes || options.dryRun
+                            ? undefined
+                            : async () => {
+                                  const ok = await p.confirm({
+                                      message: `Delete time log entry ${timeLogId.substring(0, 8)}...?`,
+                                  });
+                                  return !p.isCancel(ok) && !!ok;
+                              },
                 });
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/commands/timelog/delete.ts]: @coderabbitai Fixed in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — captured a const id after the guard so the confirm closure is a string.

File: src/azure-devops/index.ts [hunk 2]
```diff
@@ -162,6 +162,7 @@ Examples:
   # Time logging
   tools azure-devops timelog add --workitem 268935 --hours 2 --type "Development"
   tools azure-devops timelog list --workitem 268935
+  tools azure-devops timelog delete <timeLogId> --yes
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/index.ts]: @coderabbitai Fixed in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — listed delete in the Commands line and the Timelog Subcommands block.

File: src/azure-devops/lib/timelog/delete-entry.test.ts [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        const result = await deleteTimeLogEntryWithEffort(baseOpts(apis, { journalPath }));
+        expect(result.status).toBe("deleted");
+        expect(apis.calls).toContain("delete");
+        expect(apis.calls).not.toContain("update");
+
+        if (result.status === "deleted") {
+            expect(result.plan).toBeNull();
+            expect(result.effort).toBeNull();
+        }
+    });
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/lib/timelog/delete-entry.test.ts]: @coderabbitai Fixed in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — added tests for the single-id journal fallback and the unscoped project query.

File: src/azure-devops/lib/timelog/delete-entry.ts [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+            } catch (err) {
+                logger.warn(
+                    { error: err, workItemId: resolved.workItemId },
+                    "[timelog-delete] failed to read work item effort; will delete the row only"
+                );
+            }
+        } else {
+            logger.warn(`[timelog-delete] ${opts.timeLogId} not resolved; delete will skip effort restore`);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/lib/timelog/delete-entry.ts]: @coderabbitai Fixed in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — resolve throws are caught. Combined with t7, a failed resolve no longer deletes the row unless --no-effort.

File: src/azure-devops/timelog-effort.ts [hunk 5]
```diff
@@ -1,42 +1,95 @@
...
+    getWorkItem(id: number): Promise<{ rawFields?: Record<string, unknown> }>;
+    updateWorkItem(id: number, operations: JsonPatchOperation[]): Promise<unknown>;
+};
+
 /**
- * After logging time, update the work item's Remaining Work and Completed Work fields.
- * Remaining is decremented, Completed is incremented by the logged hours.
+ * Apply signed hours to Remaining / Completed.
+ * Positive minutes decrement Remaining and increment Completed.
+ * Negative minutes reverse that. Both fields floor at 0.
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/timelog-effort.ts]: @coderabbitai Fixed in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — the doc comment now says signed minutes.

File: src/azure-devops/timelog-effort.ts [hunk 6]
```diff
@@ -57,7 +110,33 @@ export async function updateWorkItemEffort(
...
+                workItemId,
+                timeLogIds: opts?.timeLogIds ?? [],
+                minutes: loggedMinutes,
+                remainingBefore: result.remainingBefore,
+                completedBefore: result.completedBefore,
+                remainingAfter: newRemaining,
+                completedAfter: newCompleted,
+            };
+
+            try {
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/timelog-effort.ts]: Partially accepted in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — a failed journal append now prints a console warning that a later delete may only restore approximately. A pending-record outbox is more machinery than this local CLI needs.

File: src/azure-devops/lib/timelog/delete-entry.ts [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+            logger.debug(
+                `[timelog-delete] resolved ${opts.timeLogId} from single-id journal #${journal.workItemId} ${journal.minutes}m`
+            );
+            return { workItemId: journal.workItemId, minutes: journal.minutes, source: "journal" };
+        }
+    }
+
+    const fromDate = queryFromDate();
+    const toDate = getTodayDate();
+    const baseQuery: TimeLogQueryParams = {
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/lib/timelog/delete-entry.ts]: Fixed in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — last-resort resolve queries with no date bound. If the entry still cannot be resolved, delete refuses unless --workitem or --no-effort is passed.

File: src/azure-devops/lib/timelog/delete-entry.ts [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    }
+
+    await opts.timeLogApi.deleteTimeLogEntry(opts.timeLogId);
+    logger.debug(`[timelog-delete] deleted ${opts.timeLogId}`);
+
+    let effort: EffortResult | null = null;
+
+    if (plan && resolved) {
+        effort = await updateWorkItemEffort(opts.devopsApi, resolved.workItemId, -resolved.minutes, {
+            timeLogIds: [opts.timeLogId],
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/lib/timelog/delete-entry.ts]: Fixed in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — rollback writes journal: false. A test asserts the original add row stays the only journal record.

File: src/azure-devops/timelog-effort-journal.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
+import { appendFileSync, existsSync, mkdirSync, readFileSync } from "node:fs";
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/timelog-effort-journal.ts]: Won't move in this PR. timelog-effort.ts already lived at the tool root before this change; the journal stays next to that writer. New delete and restore code is already under lib/timelog/. Moving the old module is a separate refactor.

File: src/azure-devops/lib/timelog/delete-entry.ts [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+    const shortId = result.timeLogId.substring(0, 8);
+    const prefix = result.status === "dry-run" ? "Would delete" : "Deleted";
+
+    out.println(`\u2714 ${prefix} time log entry ${shortId}...`);
+
+    if (opts.noEffort) {
+        out.println(pc.dim("  Work item effort left untouched (--no-effort)"));
+        return;
+    }
+
```
[the remark this answers is not in the document]
Comment (genesiscz) [reply src/azure-devops/lib/timelog/delete-entry.ts]: Fixed in [626d941ed](https://github.com/genesiscz/GenesisTools/commit/626d941ed036ce2470a071eae828effa1914973e) — an unresolved dry-run now says the row would be deleted, not that it is gone.

**Comment by genesiscz** [comment]
# Review fixes 2026-08-19 15:40 from grok-4.6

## t1 confirm closure narrowing (coderabbitai)

- **Context**: `timeLogId` is a `let` that stays `string | undefined` inside the confirm closure (TS18048).
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: accepted — capture a `const id` after the guard and use it in the callback.
- **Confidence**: 90%

## t2 help text omits delete (coderabbitai)

- **Context**: The example listed `delete`, but the Commands line and Timelog Subcommands block did not.
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: accepted — both help listings now include `delete`.
- **Confidence**: 95%

## t3 missing resolve-branch tests (coderabbitai)

- **Context**: Single-id journal fallback and unscoped project query were untested.
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: accepted — added both tests.
- **Confidence**: 90%

## t4 resolve throw aborted delete (coderabbitai)

- **Context**: `resolveTimeLogForDelete` sat outside the try, so a query 500 crashed before delete.
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: accepted, then tightened by t7 — a throw is caught; the row is not deleted unless `--no-effort`.
- **Confidence**: 90%

## t5 doc comment said hours (coderabbitai)

- **Context**: `computeEffortValues` takes minutes; the first sentence said "signed hours".
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: accepted — comment now says signed minutes.
- **Confidence**: 95%

## t6 journal append failure looks like full success (eve-bot-lovinka)

- **Context**: ADO write succeeded, journal append was swallowed, caller still got an EffortResult.
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: partially-accepted — console warning now says the journal was not written. A pending-record outbox is more machinery than this local CLI needs.
- **Confidence**: 80%

## t7 3-year query window then silent row-only delete (eve-bot-lovinka)

- **Context**: Bare-id resolve only queried 3 years; a miss still deleted the row and left effort in place.
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: accepted — last-resort query has no date bound; if still unresolved, delete refuses unless `--workitem` or `--no-effort`.
- **Confidence**: 90%

## t8 rollback journaled as a new add (eve-bot-lovinka)

- **Context**: Delete wrote a newer journal row with the same id, so a later lookup would restore the post-add values.
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: accepted — rollback uses `journal: false`; test asserts the original row stays the only record.
- **Confidence**: 95%

## t9 move effort modules under lib/timelog (eve-bot-lovinka)

- **Context**: `timelog-effort.ts` and the new journal file sit at the tool root.
- **Judge**: grok-4.6
- **Commit(s)**: none
- **Verdict**: rejected — `timelog-effort.ts` already lived at the tool root; the journal stays next to that writer. New delete/restore code is already under `lib/timelog/`. A move of the old module is a separate refactor.
- **Confidence**: 85%

## t10 dry-run said the row is gone (eve-bot-lovinka)

- **Context**: Unresolved dry-run printed "Would delete" then "The row is gone".
- **Judge**: grok-4.6
- **Commit(s)**: `626d941ed`
- **Verdict**: accepted — dry-run wording now says the row would be deleted.
- **Confidence**: 95%
